File: projects/JavaScript/benchmarks/benchmark_send_receive.js
'use strict'

const benchmark = require('benchmark')

const proto = require('../proto/proto')

class MySender1 extends proto.Sender {
  constructor () {
    super()
    this._size = 0
    this._log_size = 0
  }

  get size () {
    return this._size
  }

  get logSize () {
    return this._log_size
  }

  // noinspection JSUnusedLocalSymbols
  onSend (buffer, offset, size) {
    this._size += size
    return size
  }

  onSendLog (message) {
    this._log_size += message.length
  }
}

class MySender2 extends proto.Sender {
  constructor () {
    super()
    this._size = 0
    this._log_size = 0
  }

  get size () {
    return this._size
  }

  get logSize () {
    return this._log_size
  }

  // noinspection JSUnusedLocalSymbols
  onSend (buffer, offset, size) {
    this._size += size
    return 0
  }

  onSendLog (message) {
    this._log_size += message.length
  }
}

class MyReceiver extends proto.Receiver {
  constructor () {
    super()
    this._log_size = 0
  }

  get logSize () {
    return this._log_size
  }

  onReceive_OrderMessage (value) {} // eslint-disable-line
  onReceive_BalanceMessage (value) {} // eslint-disable-line
  onReceive_AccountMessage (value) {} // eslint-disable-line

  onReceiveLog (message) {
    this._log_size += message.length
  }
}

const suite = new benchmark.Suite('Send & Receive')

suite.on('start', function () {
  // Create a new account with some orders
  this.account = new proto.AccountMessage(new proto.Account(1, 'Test', proto.State.good, new proto.Balance('USD', 1000.0), new proto.Balance('EUR', 100.0)))
  this.account.body.orders.push(new proto.Order(1, 'EURUSD', proto.OrderSide.buy, proto.OrderType.market, 1.23456, 1000.0))
  this.account.body.orders.push(new proto.Order(2, 'EURUSD', proto.OrderSide.sell, proto.OrderType.limit, 1.0, 100.0))
  this.account.body.orders.push(new proto.Order(3, 'EURUSD', proto.OrderSide.buy, proto.OrderType.stop, 1.5, 10.0))

  this.sender1 = new MySender1()
  this.sender1.send(this.account)

  this.sender2 = new MySender2()
  this.sender2.send(this.account)

  this.receiver = new MyReceiver()
  this.receiver.receive(this.sender2.buffer)
})

suite.on('cycle', function (event) {
  console.log(String(event.target))
})

suite.add('Send', function () {
  // Serialize and send the account
  suite.sender1.send(suite.account)
})

suite.add('Send with logs', function () {
  // Enable logging
  suite.sender1.logging = true

  // Serialize and send the account
  suite.sender1.send(suite.account)
})

suite.add('Receive', function () {
  // Receive the account from the sender
  suite.receiver.receive(suite.sender2.buffer)
})

suite.add('Receive with logs', function () {
  // Enable logging
  suite.receiver.logging = true

  // Receive the account from the sender
  suite.receiver.receive(suite.sender2.buffer)
})

suite.run({ async: true })

File: projects/Swift/Tests/Benchmark/BenchmarkSendReceive.swift
//
//  BenchmarkSendReceive.swift
//

import XCTest
import ChronoxorProto
import ChronoxorFbe

fileprivate class MySender: ChronoxorProto.Sender, ChronoxorFbe.SenderListener {
    var size: Int = 0
    var logSize: Int = 0

    var logging: Bool = false

    func onSend(buffer: Data, offset: Int, size: Int) throws -> Int {
        self.size += size
        return 0
    }

    func onSendLog(message: String) {
        logSize += message.count
    }
}

fileprivate class MyReceiver: ChronoxorProto.Receiver, ChronoxorProto.ReceiverListener {
    var logSize: Int = 0
    var logging: Bool = false

    func onSendLog(message: String) {
        logSize += message.count
    }
}

class BenchmarkSendReceive: XCTestCase {
    private static var _account: AccountMessage!
    private static var _sender1: MySender!
    private static var _sender2: MySender!
    private static var _receiver: MyReceiver!

    override class func setUp() {
        super.setUp()

        _account = AccountMessage(body: Account(id: 1, name: "Test", state: .bad, wallet: ChronoxorProto.Balance(currency: "USD", amount: 1000.0), asset: ChronoxorProto.Balance(currency: "EUR", amount: 100.0), orders: []))
        _account.body.orders.append(ChronoxorProto.Order(id: 1, symbol: "EURUSD", side: ChronoxorProto.OrderSide.buy, type: OrderType.market, price: 1.23456, volume: 1000.0))
        _account.body.orders.append(ChronoxorProto.Order(id: 2, symbol: "EURUSD", side: ChronoxorProto.OrderSide.sell, type: OrderType.limit, price: 1.0, volume: 100.0))
        _account.body.orders.append(ChronoxorProto.Order(id: 3, symbol: "EURUSD", side: ChronoxorProto.OrderSide.buy, type: OrderType.stop, price: 1.5, volume: 10.0))

        _sender1 = MySender()
        _ = try? _sender1.send(value: _account)

        _sender2 = MySender()
        _ = try? _sender2.send(value: _account)

        _receiver = MyReceiver()
        try? _receiver.receive(buffer: _sender2.buffer)
    }

    func testPerformanceSend() {
        self.measure {
            for _ in 0...999 {
                // Verify the account
                _ = try? BenchmarkSendReceive._sender1.send(value: BenchmarkSendReceive._account)
            }
        }
    }

    func testPerformanceSendWithLogs() {
        // Enable logging
        BenchmarkSendReceive._sender1.logging = true

        self.measure {
            // Serialize and send the account
            _ = try? BenchmarkSendReceive._sender1.send(value: BenchmarkSendReceive._account)
        }
    }

    func testPerformanceReceive() {
        try? BenchmarkSendReceive._receiver.receive(buffer: BenchmarkSendReceive._sender2.buffer)
    }

    func testPerformanceReceiveWithLogs() {
        // Enable logging
        BenchmarkSendReceive._receiver.logging = true

        self.measure {
            // Serialize and send the account
            try? BenchmarkSendReceive._receiver.receive(buffer: BenchmarkSendReceive._sender2.buffer)
        }
    }
}
